Support tracking dependencies by branch

dmiedema opened the pull request:
In addition to `version` and all the features around that, we can also
explicitly track a remote `branch` that will be fetched/updated when
`modulo update` is run.

Add `make test` command
Lets make running tests easier by adding a Make command for it

dmiedema committed:
Add `make test` command

Lets make running tests easier by adding a Make command for it

dmiedema committed:
Merge branch 'master' of https://github.com/modulo-dm/modulo

dmiedema committed:
Support tracking dependencies by branch

In addition to `version` and all the features around that, we can also
explicitly track a remote `branch` that will be fetched/updated when
`modulo update` is run.

bsneed reviewed (changes requested):
Not really what was discussed in slack yesterday as a solution.

Command should be:

--unmanaged <branch/tag/hash/nothing>

Where, a branch/tag/hash can be specified, or simply nothing.  No need to store it other than for checkout purposes.

dmiedema committed:
Update for feedback & remove `--branch` option

Make for the `--unmanaged` flag on `add` to have an optional argument of
a commit or a branch. However this lead to complications around other
flags being passed such as `add blah --unmanaged -u -v`.

To resolve that issue, the properties on `Option` were move to `public`
so the add command could try its best to still apply arguments when
they were inadvertantly swallowed by the now `--unmanaged` optional
value.

Also the `unmanagedValue` is tracked in the `.modulo` file like `branch`
was but is more generic supporting an entire commit hash or a branch
name

dmiedema commented:
Expanding on the message in 
https://github.com/modulo-dm/modulo/pull/16/commits/125e682d65d59a99bfbb848684c33b0951026872

The main weirdness is that it doesn't look like ELCLI supports an optional value. Which makes sense, that's a unique use case. It could be extended to support just that but for now, the fastest change I could do was to make the properties on `Option` public and attempt to reapply the swallowed flag fir this one particular use case.

dmiedema committed:
Cleanup hack to process flags in add command

Rather than have the add command hack in support for optional values to
flags make it a core feature of `ELCLI`. That seems like the proper
place to handle this situation and also doesn't break existing
functionality/assumptions in all the other uses/tests.

johnclayton commented:
I had an understanding of what this was going to be more akin to what @bsneed had. The `--unmanaged <hash|branch|nothing>` option would not track the value going forward but just do the initial checkout. 

I also think the interface should be normalized. The `add` command command should be able to manage a version|hash|branch through the same interface, and unmanage with a version as well. 

So:

```sh
$ modulo add --version <version|hash|branch>
$ modulo add --unmanaged <version|hash|branch|nothing>
```

Maybe `--version` is not the right place to put those other two value types but it's nice to be able to just track a branch during development, and a specific hash can really help when encountering a bug or something that requires a temporary pin to a specific commit.

cgarvin647 commented on ModuloKit/Specs/DependencySpec.swift:
```diff
@@ -17,6 +17,9 @@ public struct DependencySpec {
     var version: SemverRange?
+    /// Optional unmanaged property to track
+    /// such as a branch name, commit hash, or nothing
+    var unmanagedValue: String?
```
Is unmanagedValue the right name for this? Maybe we could use `unmanagedBranchOrHash`, `unmanagedCheckout`. Honestly I'm not so sure I agree with the `--unmanaged` name option. It isn't unmanaging anything, it's more not using a version. `unversioned`?

cgarvin647 reviewed (commented):
I'd like to see some clearer names being used. I like the direction of this.

dmiedema commented:
Yeah, `--unmanaged` is probably not the flag to attach this to, and `unmanagedValue` really isn't the best naming choice. 

Maybe `--unmanaged` remains exactly as is but an additional flag is added along side `--version`

Maybe something like `--tracked` or `--managed` and `{tracked|managed}BranchOrHash` for the JSON key? That kind of makes sense to me and is far more generic than my initial `--branch` flag

That might be more explicit and easier to understand also. It would render the changes in ECLI unnecessary (although they may be useful on their own? I'm not sure of all of its use cases) as well.

bsneed commented:
Nah, these comments are getting away from the point.

Unmanaged = You manage it yourself.  Checkout whatever branch/hash/tag you desire.  Optionally, it will do the initial checkout for you, nothing more.

You want it managed, you use a version.  The end.  No branches/hashes/non-semver tags, etc.

@dmiedema hit me up on slack if this doesn't clear it up.
